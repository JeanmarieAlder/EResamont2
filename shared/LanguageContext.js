import { createContext } from "react";

export const LanguageContext = createContext();
// 1 is French
// 2 is Italian
// 3 is English
// 4 is Deutsch
