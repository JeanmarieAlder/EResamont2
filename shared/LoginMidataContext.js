import React, { useState, useEffect, useContext } from "react";
import { ToastAndroid } from "react-native";
import * as AppAuth from "expo-app-auth";
import * as SecureStore from "expo-secure-store";
import requestGetMidata from "../utils/requestGetMidata";

export const Auth2Context = React.createContext();
export const useAuth2 = () => useContext(Auth2Context);
export const Auth2Provider = ({ children }) => {
  const StorageKey = "EResamont2MidataOAuthKey";
  const config = {
    issuer: "https://test.midata.coop/fhir/metadata",
    clientId: "eresamont2-test",
    serviceConfiguration: {
      authorizationEndpoint: "https://test.midata.coop/authservice",
      tokenEndpoint: "https://test.midata.coop/v1/token",
      revocationEndpoint: "https://test.midata.coop/v1/token"
    },
    redirectUrl: AppAuth.OAuthRedirect + ":/oauthredirect"
  };

  const [authState, setAuthState] = useState(null);
  const [userName, setUserName] = useState(null);

  useEffect(() => {
    console.log("LoginMidataContext: useEffect()");
    (async () => {
      let cachedAuth = await getCachedAuthAsync();
      if (cachedAuth && !authState) {
        setAuthState(cachedAuth);
      }
    })();
  }, []);

  const signInAsync = async () => {
    console.log("signInAsync()");
    let authState = await AppAuth.authAsync(config);
    await cacheAuthAsync(authState);
    console.log("signInAsync", authState);
    setAuthState(authState);
    return authState;
  };

  const signOutAsync = async ({ accessToken }) => {
    try {
      await AppAuth.revokeAsync(config, {
        token: accessToken,
        isClientIdProvided: true
      });
      await SecureStore.deleteItemAsync(StorageKey);
      console.log("sign out completed");
      ToastAndroid.show("Sign out completed", ToastAndroid.SHORT);
      setAuthState(null);
      return null;
    } catch (e) {
      console.error(`Failed to revoke token: ${e.message}`);
    }
  };

  const cacheAuthAsync = async authState => {
    return await SecureStore.setItemAsync(
      StorageKey,
      JSON.stringify(authState)
    );
  };

  const getCachedAuthAsync = async () => {
    let value = await SecureStore.getItemAsync(StorageKey);
    let authState = JSON.parse(value);
    console.log("getCachedAuthAsync", authState);
    if (authState) {
      if (checkIfTokenExpired(authState)) {
        return refreshAuthAsync(authState);
      } else {
        return authState;
      }
    }
    return null;
  };

  const checkIfTokenExpired = ({ accessTokenExpirationDate }) => {
    return new Date(accessTokenExpirationDate) < new Date();
  };

  const refreshAuthAsync = async ({ refreshToken }) => {
    let authState = await AppAuth.refreshAsync(config, refreshToken);
    console.log("refreshAuth", authState);
    await cacheAuthAsync(authState);
    return authState;
  };

  const getLoggedUserName = async () => {
    if (authState) {
      let response = await requestGetMidata(
        "https://test.midata.coop/fhir/Patient/" +
          authState.additionalParameters.patient,
        authState,
        signInAsync
      );
      if (response) {
        setUserName(response.name[0].given[0]);
        return true;
      }
    }
    return false;
  };

  return (
    <Auth2Context.Provider
      value={{
        authState,
        userName,
        signInAsync,
        signOutAsync,
        cacheAuthAsync,
        getCachedAuthAsync,
        checkIfTokenExpired,
        refreshAuthAsync,
        getLoggedUserName
      }}
    >
      {children}
    </Auth2Context.Provider>
  );
};
