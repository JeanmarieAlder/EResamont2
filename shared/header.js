import React from "react";
import { StyleSheet, Text, View } from "react-native";
import { MaterialIcons } from "@expo/vector-icons";

export default function Header({ title, navigation }) {
  const openMenu = () => {
    navigation.openDrawer();
  };

  return (
    <View style={styles.header}>
      <MaterialIcons
        name="menu"
        size={28}
        onPress={openMenu}
        style={styles.icon}
      />
      <View>
        <Text style={styles.headerText}>{title}</Text>
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  header: {
    flexDirection: "row",
    justifyContent: "flex-end",
    flex: 1
  },
  headerText: {
    fontWeight: "bold",
    fontSize: 20,
    color: "#fff",
    alignItems: "flex-end",
    letterSpacing: 1,
    marginLeft: 10
  },
  icon: { color: "#fff" }
});
