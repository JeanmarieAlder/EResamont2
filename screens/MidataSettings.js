import React, { useContext } from "react";
import {
  StyleSheet,
  View,
  Text,
  Image,
  ToastAndroid,
  Alert,
  ScrollView
} from "react-native";
import { globalStyles } from "../styles/global";
import { LanguageContext } from "../shared/LanguageContext";
import ButtonView from "../components/ButtonView";
import { useAuth2 } from "../shared/LoginMidataContext";
import requestGetMidata from "../utils/requestGetMidata";
import storage from "../utils/storage";
export default function MidataSettings({ navigation }) {
  const { language, setLanguage } = useContext(LanguageContext);
  const {
    authState,
    userName,
    signInAsync,
    signOutAsync,
    getLoggedUserName
  } = useAuth2();

  let logInPressed = async () => {
    if (authState) {
      ToastAndroid.show("Allready logged in", ToastAndroid.SHORT);
    } else {
      signInAsync();
    }
  };

  let logOutPressed = async () => {
    if (authState) {
      signOutAsync(authState);
    } else {
      ToastAndroid.show("Already logged off", ToastAndroid.SHORT);
    }
  };
  const confirmClearScoreClick = async idQuizz => {
    let result = await storage.getQuizScore(idQuizz);
    if (result) {
      Alert.alert(
        "Confirm local score deletion",
        "Current scores are: " + JSON.stringify(result),
        [
          { text: "YES", onPress: () => storage.deleteScoreData(idQuizz) },
          {
            text: "NO",
            style: "cancel"
          }
        ],
        { cancelable: true }
      );
    }
  };
  return (
    <View style={globalStyles.container}>
      <Image
        source={require("../assets/images/logo_midata.png")}
        style={localStyles.logoImage}
      />
      <Text>
        Midata is a citizen owned, non-profit and secure platform to share
        medical data for medical research. You decide the data you want to
        share, researchers will use them to create new treatments for everyone.
      </Text>
      {authState && getLoggedUserName() ? (
        <Text style={localStyles.loginStatusText}>Welcome {userName}!</Text>
      ) : (
        <Text style={localStyles.loginStatusText}>Sign in to use Midata</Text>
      )}
      <View style={localStyles.midataContainer}>
        <ScrollView contentContainerStyle={localStyles.scrollViewMain}>
          <ButtonView
            value="Log in / Sign up"
            style={{
              ...globalStyles.midataButton,
              backgroundColor: "darkblue"
            }}
            onPress={async () => {
              logInPressed();
            }}
          />
          <ButtonView
            value="My data"
            style={{
              ...globalStyles.midataButton,
              backgroundColor: "darkblue"
            }}
            onPress={() => {
              navigation.push("MyData");
            }}
          />
          <ButtonView
            value="Clear local lake louise data (debug)"
            style={{
              ...globalStyles.midataButton,
              backgroundColor: "darkblue"
            }}
            onPress={() => {
              confirmClearScoreClick(95);
            }}
          />
          <ButtonView
            value="Clear local oxygen data (debug)"
            style={{
              ...globalStyles.midataButton,
              backgroundColor: "darkblue"
            }}
            onPress={() => {
              confirmClearScoreClick(100);
            }}
          />
          <ButtonView
            value="Login status (debug console)"
            style={{
              ...globalStyles.midataButton,
              backgroundColor: "darkblue"
            }}
            onPress={() => console.log(authState)}
          />
          <ButtonView
            value="Get my posted data (debug)"
            style={{
              ...globalStyles.midataButton,
              backgroundColor: "darkblue"
            }}
            onPress={async () => {
              if (authState) {
                let response = await requestGetMidata(
                  "https://test.midata.coop/fhir/QuestionnaireResponse/",
                  authState,
                  signInAsync
                );
                console.log(response);
              } else {
                signInAsync();
              }
            }}
          />
          <ButtonView
            value="Sign out"
            style={{ ...globalStyles.midataButton, backgroundColor: "grey" }}
            onPress={async () => {
              logOutPressed();
            }}
          />
        </ScrollView>
      </View>
    </View>
  );
}

const localStyles = StyleSheet.create({
  logoImage: {
    width: "100%",
    resizeMode: "contain",
    alignSelf: "center"
  },
  midataContainer: {
    flex: 1
  },
  loginStatusText: {
    fontWeight: "bold",
    alignSelf: "center",
    fontSize: 15
  },
  buttonContainerMain: { flex: 3 },
  scrollViewMain: {
    flexGrow: 1,
    justifyContent: "flex-start",
    alignItems: "center"
  }
});
