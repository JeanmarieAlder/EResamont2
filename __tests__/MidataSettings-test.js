import * as React from "react";
import { render, fireEvent } from "react-native-testing-library";
import renderer from "react-test-renderer";
import { LanguageContext } from "../shared/LanguageContext";
import { Auth2Provider } from "../shared/LoginMidataContext";
import MidataSettings from "../screens/MidataSettings";
import NavigationTestUtils from "react-navigation/NavigationTestUtils";
import { Auth2Context } from "../shared/LoginMidataContext";
import { ToastAndroid } from "react-native";

let language = 1;
console.log = jest.fn();
ToastAndroid.show = jest.fn();

const fakeNavigation = {
  push: (location) => {
    fakePush(location);
  },
};

let fakePush = jest.fn();

const FakeAuth2ProviderOnline = ({ children }) => {
  return (
    <Auth2Context.Provider
      value={{
        authState: true,
        getLoggedUserName: () => {
          return "testuser";
        },
        signOutAsync: () => {
          fakeSignOutAsync();
        },
      }}
    >
      {children}
    </Auth2Context.Provider>
  );
};
const FakeAuth2ProviderOffline = ({ children }) => {
  return (
    <Auth2Context.Provider
      value={{
        authState: false,
        getLoggedUserName: () => {
          return "testuser";
        },
        signInAsync: () => {
          fakeSignInAsync();
        },
        signOutAsync: () => {
          fakeSignOutAsync();
        },
      }}
    >
      {children}
    </Auth2Context.Provider>
  );
};
let fakeSignOutAsync = jest.fn();
let fakeSignInAsync = jest.fn();

describe("MidataSettings", () => {
  beforeEach(() => {
    NavigationTestUtils.resetInternalState();
    //reset mocks and counters
    jest.clearAllMocks();
  });
  it("Renders correctly", () => {
    let component = renderer
      .create(
        <Auth2Provider>
          <LanguageContext.Provider value={{ language }}>
            <MidataSettings />
          </LanguageContext.Provider>
        </Auth2Provider>
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  describe("LogInPressed", () => {
    it("says user is allready logged in if so", async () => {
      const { getByTestId } = render(
        <FakeAuth2ProviderOnline>
          <LanguageContext.Provider value={{ language }}>
            <MidataSettings />
          </LanguageContext.Provider>
        </FakeAuth2ProviderOnline>
      );
      const element = getByTestId("button-view-Log in / Sign up");
      await fireEvent.press(element);
      expect(ToastAndroid.show).toHaveBeenCalledTimes(1);
      expect(ToastAndroid.show).toHaveBeenCalledWith(
        "Allready logged in",
        ToastAndroid.SHORT
      );
    });
    it("proceeds to login if not logged in", async () => {
      const { getByTestId } = render(
        <FakeAuth2ProviderOffline>
          <LanguageContext.Provider value={{ language }}>
            <MidataSettings />
          </LanguageContext.Provider>
        </FakeAuth2ProviderOffline>
      );
      const element = getByTestId("button-view-Log in / Sign up");
      await fireEvent.press(element);
      expect(fakeSignInAsync).toHaveBeenCalledTimes(1);
    });
  });
  describe("LogOutPressed", () => {
    it("says user is allready logged off if so", async () => {
      const { getByTestId } = render(
        <FakeAuth2ProviderOffline>
          <LanguageContext.Provider value={{ language }}>
            <MidataSettings />
          </LanguageContext.Provider>
        </FakeAuth2ProviderOffline>
      );
      const element = getByTestId("button-view-Sign out");
      await fireEvent.press(element);
      expect(ToastAndroid.show).toHaveBeenCalledWith(
        "Already logged off",
        ToastAndroid.SHORT
      );
      expect(ToastAndroid.show).toHaveBeenCalledTimes(1);
    });

    it("proceeds to logout if logged in", async () => {
      const { getByTestId } = render(
        <FakeAuth2ProviderOnline>
          <LanguageContext.Provider value={{ language }}>
            <MidataSettings />
          </LanguageContext.Provider>
        </FakeAuth2ProviderOnline>
      );
      const element = getByTestId("button-view-Sign out");
      await fireEvent.press(element);
      expect(fakeSignOutAsync).toHaveBeenCalledTimes(1);
    });
  });
  describe("MyData button", () => {
    it("navigates to my data page", async () => {
      const { getByTestId } = render(
        <FakeAuth2ProviderOnline>
          <LanguageContext.Provider value={{ language }}>
            <MidataSettings navigation={fakeNavigation} />
          </LanguageContext.Provider>
        </FakeAuth2ProviderOnline>
      );
      const element = getByTestId("button-view-My data");
      await fireEvent.press(element);
      expect(fakePush).toHaveBeenCalledWith("MyData");
      expect(fakePush).toHaveBeenCalledTimes(1);
    });
  });
});
