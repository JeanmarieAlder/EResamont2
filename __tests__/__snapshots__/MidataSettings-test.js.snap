// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MidataSettings Renders correctly 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
>
  <Image
    source={1}
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": "100%",
          "width": "100%",
        },
        undefined,
      ]
    }
  />
  <View
    style={
      Object {
        "flex": 1,
        "padding": 20,
      }
    }
  >
    <Image
      source={1}
      style={
        Object {
          "alignSelf": "center",
          "resizeMode": "contain",
          "width": "100%",
        }
      }
    />
    <Text
      style={
        Object {
          "backgroundColor": "rgba(255,255,255, 0.7)",
          "borderRadius": 5,
          "padding": 5,
        }
      }
    >
      Midata is a citizen owned, non-profit and secure platform to share medical data for medical research. You decide the data you want to share, researchers will use them to create new treatments for everyone.
    </Text>
    <Text
      style={
        Object {
          "alignSelf": "center",
          "backgroundColor": "rgba(255,255,255, 0.5)",
          "fontSize": 15,
          "fontWeight": "bold",
          "marginTop": 3,
          "padding": 5,
        }
      }
    >
      Please sign in to use Midata
    </Text>
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <RCTScrollView
        contentContainerStyle={
          Object {
            "alignItems": "center",
            "flexGrow": 1,
            "justifyContent": "flex-start",
          }
        }
      >
        <View>
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "darkblue",
                "borderRadius": 10,
                "height": 40,
                "justifyContent": "center",
                "marginTop": 10,
                "opacity": 1,
                "padding": 10,
                "width": "80%",
              }
            }
            testID="button-view-Sign in to Midata"
          >
            <Text
              style={
                Object {
                  "color": "#ffffff",
                  "fontFamily": "EuphemiaUCAS-Bold",
                  "textAlign": "center",
                  "textAlignVertical": "center",
                }
              }
            >
              Sign in to Midata
            </Text>
          </View>
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "darkblue",
                "borderRadius": 10,
                "height": 40,
                "justifyContent": "center",
                "marginTop": 10,
                "opacity": 1,
                "padding": 10,
                "width": "80%",
              }
            }
            testID="button-view-My data"
          >
            <Text
              style={
                Object {
                  "color": "#ffffff",
                  "fontFamily": "EuphemiaUCAS-Bold",
                  "textAlign": "center",
                  "textAlignVertical": "center",
                }
              }
            >
              My data
            </Text>
          </View>
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "grey",
                "borderRadius": 10,
                "height": 40,
                "justifyContent": "center",
                "marginTop": 10,
                "opacity": 1,
                "padding": 10,
                "width": "80%",
              }
            }
            testID="button-view-Sign out"
          >
            <Text
              style={
                Object {
                  "color": "#ffffff",
                  "fontFamily": "EuphemiaUCAS-Bold",
                  "textAlign": "center",
                  "textAlignVertical": "center",
                }
              }
            >
              Sign out
            </Text>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
</View>
`;
